plutus-ledger/src/Ledger/Tx.hs

```diff
@@ -101,13 +103,14 @@ type PrivateKey = Crypto.XPrv
 --
 -- TODO Add 'Either DatumHash Datum' field for 'PublicKeyChainIndexTxOut'.
 data ChainIndexTxOut =
-    PublicKeyChainIndexTxOut { _ciTxOutAddress :: Address
-                             , _ciTxOutValue   :: V1.Value
+    PublicKeyChainIndexTxOut { _ciTxOutAddress        :: Address
+                             , _ciTxOutValue          :: V1.Value
+                             , _ciTxOutDatumPublicKey :: V2.OutputDatum
                              }
-  | ScriptChainIndexTxOut { _ciTxOutAddress   :: Address
-                          , _ciTxOutValue     :: V1.Value
-                          , _ciTxOutDatum     :: Either V1.DatumHash V1.Datum
-                          , _ciTxOutValidator :: Either V1.ValidatorHash V1.Validator
+  | ScriptChainIndexTxOut { _ciTxOutAddress     :: Address
+                          , _ciTxOutValue       :: V1.Value
+                          , _ciTxOutDatumScript :: Either V1.DatumHash V1.Datum
```

Comment:
I split `_ciTxOutDatum` into `_ciTxOutDatumPublicKey` and `_ciTxOutDatumScript` because the types are now different. If you have a better idea let me know.